implementation/src/main/java/io/smallrye/graphql/execution/LRUCache.java

```diff
@@ -0,0 +1,100 @@
+package io.smallrye.graphql.execution;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+public class LRUCache<K, V> {
+
+    private final int maxSize;
+    private final Map<K, Entry<V>> cache = new ConcurrentHashMap<>();
+    private Entry<V> start;
+    private Entry<V> end;
+
+    LRUCache(int maxSize) {
+        this.maxSize = maxSize;
+    }
+
+    V get(K key) {
+        Entry<V> entry = cache.computeIfPresent(key, this::moveEntryToStart);
+        return entry == null ? null : entry.value;
+    }
+
+    V put(K key, V value) {
+        Entry<V> entry = cache.get(key);
+        V oldValue = null;
+        if (entry == null) {
+            entry = cachePutNew(key, value);
+        } else {
+            entry = removeEntry(entry);
+            oldValue = entry.value;
+            entry.value = value;
+        }
+        entry = addToStart(entry);
+        return oldValue;
+    }
+
+    V computeIfAbsent(K key, Function<K, V> valueFunction) {
+        Entry<V> entry = cache.get(key);
+        if (entry == null) {
+            entry = cachePutNew(key, valueFunction.apply(key));
+        } else {
+            entry = removeEntry(entry);
+        }
+        entry = addToStart(entry);
+        return entry.value;
+    }
+
+    private Entry<V> cachePutNew(K key, V value) {
+        Entry<V> entry = new Entry<V>(key, value);
+        cache.put(key, entry);
+        synchronized (cache) {
+            if (cache.size() > maxSize) {
+                cache.remove(end.key);
+                removeEntry(end);
+            }
+        }
+        return entry;
+    }
+
+    private synchronized Entry<V> moveEntryToStart(K key, Entry<V> entry) {
+        return addToStart(removeEntry(entry));
+    }
+
+    private synchronized Entry<V> removeEntry(Entry<V> entry) {
+        if (entry.left != null) {
+            entry.left.right = entry.right;
+        }
```

Comment:
Do you need an else here?
```
else {
    start = entry.right;
}
```

Reply:
Maybe for completeness, but it shouldn't be required.  Every time that we remove and entry it is later added to the start of the list, so the `start` field is reset. The one exception to that is when the cache is full, then we always remove from the end - so the only time we would need to reset `start` in this method is if the max cache size were set to 0 or lower - but in that case, we would never use `start` or `end`.  I have no problem adding the else block, but I don't think it will ever be used.

Reply:
If remove is called and it is the only thing in the cache, then start should be set to null.  Without the else, it will be left set to the one entry that was in the cache.